### PR TITLE
Pa02

### PR DESCRIPTION
**Pull request by Johanstab**
Did the project according to description. Partner: Sabina Langås. 
Feedback is appreciated.

### CODE REVIEW

**Review by sababina — APPROVED** [review]

File: src/pa02/chutes_simulation.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        if self.fell_down:
+            extra = self.plus_step
+        else:
+            extra = 0
+
+        roll = random.randint(1, 6)
+
+        self.position += roll + extra
+        self.position += self.board.position_adjustment(self.position)
+        self.number_of_moves += 1
```
Comment (huynhngoc) [inline src/pa02/chutes_simulation.py:135]: `self.fell_down` will never become True in your code. --> The player will never take the extra steps

File: src/pa02/chutes_simulation.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+        if self.climbed:
+            extra = self.minus_step
+        else:
+            extra = 0
+
+        roll = random.randint(1, 6)
+
+        self.position += roll + extra
+        self.position += self.board.position_adjustment(self.position)
+        self.number_of_moves += 1
```
Comment (huynhngoc) [inline src/pa02/chutes_simulation.py:163]: Similarly with `self.climbed`

File: src/pa02/chutes_simulation.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+        randomize_players: If the players should be in randomized order
+        """
+        if board is None:
+            self.board = Board()
+        else:
+            self.board = board
+
+        self.player_types = frozenset(c.__name__ for c in player_field)
+        self.players = player_field
+        self.seed = random.seed(seed)
```
Comment (huynhngoc) [inline src/pa02/chutes_simulation.py:188]: Don't need to store `self.seed` here.

File: src/pa02/chutes_simulation.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        self.randomize_players = randomize_players
+        self.results = []
+
+        if self.randomize_players is True:
+            random.shuffle(self.players)
```
Comment (huynhngoc) [inline src/pa02/chutes_simulation.py:193]: This should be done at the beginning of every single_game

File: src/pa02/test_pa02.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+        board = cs.Board()
+        player = cs.ResilientPlayer(board)
+
+        for step in range(50):
+            player.move()
+            assert player.position not in board.snakes_and_ladders.keys()
+
+
+class TestSimulation:
```
Comment (huynhngoc) [inline src/pa02/test_pa02.py:149]: This is not how we handle randomness in real-life. And you used the result from your code to create tests, which makes these tests meaningless.

Comment (huynhngoc) [reply src/pa02/test_pa02.py:149]: There are other things you can test in Simulation:
- Single_game: make the board with ladders: 1 --> 90, 2-->90, 3 --> 90, 4 --> 90, 5 -->90, 6-->90. Then the first player will always win.
- Results: length of the return list should be the same with number of simulation. The first element in each tuple in that list should always greater than 0, and the second element should belongs to the list of different players.
- winners: the sum of all values in the dict should be equal with number of simulation.
- duration: the sum of the length of each value in the dict should be equal with number of simulation.

**Review by huynhngoc — APPROVED** [review]
Approved.
Please check the comments.

**Pull request merged**